journal/src/main/java/io/camunda/zeebe/journal/file/SegmentedJournal.java

```diff
@@ -521,6 +520,29 @@ private boolean handleSegmentCorruption(
     return Arrays.asList(files);
   }
 
+  private void deleteDeferredFiles() {
+    final File[] files =
+        directory.listFiles(
+            file -> file.isFile() && JournalSegmentFile.isDeletedSegmentFile(name, file.getName()));
+
+    if (files == null) {
+      throw new IllegalStateException(
+          String.format(
+              "Could not list files in directory '%s'. Either the path doesn't point to a directory or an I/O error occurred.",
+              directory));
+    }
+
+    Arrays.stream(files)
+        .forEach(
+            file -> {
+              try {
+                Files.deleteIfExists(file.toPath());
+              } catch (final IOException e) {
+                log.warn("Could not delete file {} which is marked for deletion", file, e);
+              }
+            });
```

Comment:
:wrench: I think you can simplify this if you just convert `directory` to a path initially, then use a directory stream. Then `files` is never null, and you can filter on the stream and then in `forEach` call the delete code.